{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "97d39060",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>pop1_A</th>\n",
       "      <th>pop1_T</th>\n",
       "      <th>pop1_G</th>\n",
       "      <th>pop1_C</th>\n",
       "      <th>pop2_A</th>\n",
       "      <th>pop2_T</th>\n",
       "      <th>pop2_G</th>\n",
       "      <th>pop2_C</th>\n",
       "      <th>pop3_A</th>\n",
       "      <th>pop3_T</th>\n",
       "      <th>pop3_G</th>\n",
       "      <th>pop3_C</th>\n",
       "      <th>pop4_A</th>\n",
       "      <th>pop4_T</th>\n",
       "      <th>pop4_G</th>\n",
       "      <th>pop4_C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG916831.1</td>\n",
       "      <td>14867521</td>\n",
       "      <td>0</td>\n",
       "      <td>254</td>\n",
       "      <td>0</td>\n",
       "      <td>99</td>\n",
       "      <td>0</td>\n",
       "      <td>346</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>244</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HG916831.1</td>\n",
       "      <td>14867535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>473</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>533</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>407</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>381</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HG916831.1</td>\n",
       "      <td>14867549</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>677</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>488</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HG916831.1</td>\n",
       "      <td>14867557</td>\n",
       "      <td>0</td>\n",
       "      <td>569</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>706</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>486</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>HG916831.1</td>\n",
       "      <td>14867568</td>\n",
       "      <td>0</td>\n",
       "      <td>589</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>736</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>494</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55944</th>\n",
       "      <td>HG916850.1</td>\n",
       "      <td>9679382</td>\n",
       "      <td>0</td>\n",
       "      <td>730</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>469</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>430</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>552</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55954</th>\n",
       "      <td>HG916850.1</td>\n",
       "      <td>9679392</td>\n",
       "      <td>656</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>399</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>521</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55967</th>\n",
       "      <td>HG916850.1</td>\n",
       "      <td>9679405</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>550</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>393</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>325</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55968</th>\n",
       "      <td>HG916850.1</td>\n",
       "      <td>9679406</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>539</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>391</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>321</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>405</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55974</th>\n",
       "      <td>HG916850.1</td>\n",
       "      <td>9679412</td>\n",
       "      <td>0</td>\n",
       "      <td>434</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>348</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>254</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>330</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6592 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome  position  pop1_A  pop1_T  pop1_G  pop1_C  pop2_A  pop2_T  \\\n",
       "5      HG916831.1  14867521       0     254       0      99       0     346   \n",
       "19     HG916831.1  14867535       0       0     473       0       0      11   \n",
       "33     HG916831.1  14867549       0       0     547       0      11       0   \n",
       "41     HG916831.1  14867557       0     569       0       8       0     706   \n",
       "52     HG916831.1  14867568       0     589       0       0       0     736   \n",
       "...           ...       ...     ...     ...     ...     ...     ...     ...   \n",
       "55944  HG916850.1   9679382       0     730       0       0       0     469   \n",
       "55954  HG916850.1   9679392     656       0      14       0     438       0   \n",
       "55967  HG916850.1   9679405      10       0       0     550       0       0   \n",
       "55968  HG916850.1   9679406       0       0     539      15       0       0   \n",
       "55974  HG916850.1   9679412       0     434       0      24       0     348   \n",
       "\n",
       "       pop2_G  pop2_C  pop3_A  pop3_T  pop3_G  pop3_C  pop4_A  pop4_T  pop4_G  \\\n",
       "5           0      83       0     244       0      82       0     237       0   \n",
       "19        533       0       0       0     407       0       0       5     381   \n",
       "33        677       0       5       0     488       0      11       0     460   \n",
       "41          0      15       0     486       0      22       0     467       0   \n",
       "52          0       0       7     494       0       0       0     496       0   \n",
       "...       ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "55944       0      11       0     430       0       0       0     552       0   \n",
       "55954       0       0     399       0       0       0     521       0       0   \n",
       "55967       0     393       0       0       0     325       6       0       0   \n",
       "55968     391       0       0       0     321       0       0       0     405   \n",
       "55974       0      22       0     254       0      34       0     330       0   \n",
       "\n",
       "       pop4_C  \n",
       "5          73  \n",
       "19          0  \n",
       "33          0  \n",
       "41         22  \n",
       "52          0  \n",
       "...       ...  \n",
       "55944       9  \n",
       "55954       0  \n",
       "55967     411  \n",
       "55968       7  \n",
       "55974      21  \n",
       "\n",
       "[6592 rows x 18 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the pandas library and read the Genotypes_capstone.csv file, using ';' as the delimiter.\n",
    "import pandas as pd\n",
    "genotypes = pd.read_csv(\"Genotypes_capstone.csv\", sep=\";\")\n",
    "\n",
    "# Define a function that takes a list of genotype values\n",
    "# and returns a list of True/False indicating if each value is greater than 0.\n",
    "def find_present_nucleotides(nucs_in_population):\n",
    "    return [x > 0 for x in (nucs_in_population)]\n",
    "\n",
    "def is_biallelic(row):\n",
    "    # Call the 'find_present_nucleotides' function on specific columns for every population group\n",
    "    # to create lists of True/False values.\n",
    "    # Combine the True/False lists using the 'and' operator to create a total presence list.\n",
    "    nuc_presence_total = [any(items) for items in \n",
    "                          zip(find_present_nucleotides(row[2:6]), \n",
    "                              find_present_nucleotides(row[6:10]), \n",
    "                              find_present_nucleotides(row[10:14]), \n",
    "                              find_present_nucleotides(row[14:18]))]\n",
    "    return sum(nuc_presence_total) == 2\n",
    "\n",
    "\n",
    "# Filter the genotypes to include only biallelic SNPs.\n",
    "filtered_genotypes = genotypes[genotypes.apply(is_biallelic, axis=1)]\n",
    "filtered_genotypes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "cee1c438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of biallelic SNPs: 6592\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of biallelic SNPs: {filtered_genotypes.shape[0]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
